src/lib.rs

```diff
@@ -298,18 +298,27 @@ where
                 maybe_tt = self.activity_heartbeat_manager_handle.next_pending_cancel() => {
                     // Issue cancellations for anything we noticed was cancelled during heartbeating
                     if let Some(task_token) = maybe_tt {
+                        // Remove the outstanding activity task, since we are issuing a cancel.
+                        // If we don't, lang side might (reasonably) not complete any outstanding
+                        // ATs, because it was cancelled, but by not doing so it will leave those
+                        // tasks outstanding forever and eventually we'll stall when we hit the
+                        // outstanding limit. Lang can still complete them, and they will be sent
+                        // to server, where they may or may not actually be found.
+
                         // It's possible that activity has been completed and we no longer have an
                         // outstanding activity task. This is fine because it means that we no
                         // longer need to cancel this activity, so we'll just ignore such orphaned
                         // cancellation.
-                        if let Some(details) = self.outstanding_activity_tasks.get(&task_token) {
-                            return Ok(ActivityTask::cancel_from_ids(
-                                task_token,
-                                details.activity_id.clone(),
-                            ));
+                        if let Some((_, details)) =
+                            self.outstanding_activity_tasks.remove(&task_token) {
```

Comment:
I don't understand why is cleanup during completion is insufficient. No matter how cancellation is dispatched activity should at the end be completed by lang, which is when we should remove from this map. Otherwise we lose track of what is actually in progress on the lang side.

Reply:
Yeah, not doing this.